core/server/src/main/java/alluxio/worker/block/BlockWorker.java

```diff
@@ -268,7 +268,14 @@ public void accessBlock(long sessionId, long blockId) throws BlockDoesNotExistEx
   public void commitBlock(long sessionId, long blockId)
       throws BlockAlreadyExistsException, BlockDoesNotExistException, InvalidWorkerStateException,
       IOException, WorkerOutOfSpaceException {
-    mBlockStore.commitBlock(sessionId, blockId);
+    // NOTE: this may be invoked multiple times due to retry on client side.
+    // TODO(binfan): find a better way to handle retry logic
+    try {
+      mBlockStore.commitBlock(sessionId, blockId);
+    } catch (BlockAlreadyExistsException e) {
+      LOG.debug("Block {} has been in block store, this could be a retry due to master-side RPC "
```

Comment:
handling this case well is actually the goal: `safely continue while two threads are trying to commit the same block at the same time`, we should have a good story in the next release. This fix is a workaround to rescue the case one client thread trying twice (first failed) by retrying and throwing a more meaningful exception. The case two threads writing to the same block is neither handled well before or after this PR.